File: java/benchmarks/src/test/java/javabushka/client/Benchmarking.java
package javabushka.client;

import javabushka.client.LatencyResults;
import javabushka.client.ChosenAction;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.Collections;

public class Benchmarking {
    static final double PROB_GET = 0.8;
    static final double PROB_GET_EXISTING_KEY = 0.8;
    static final int SIZE_GET_KEYSPACE = 3750000;
    static final int SIZE_SET_KEYSPACE = 3000000;

    private static ChosenAction chooseAction() {
        if (Math.random() > PROB_GET) {
            return ChosenAction.SET;
        }
        if (Math.random() > PROB_GET_EXISTING_KEY) {
            return ChosenAction.GET_NON_EXISTING;
        }
        return ChosenAction.GET_EXISTING;
    }

    public static String generateKeyGet() {
        int range = SIZE_GET_KEYSPACE - SIZE_SET_KEYSPACE;
        return Math.floor(Math.random() * range + SIZE_SET_KEYSPACE + 1) + "";
    }
   
    public static String generateKeySet() {
        return (Math.floor(Math.random() * SIZE_SET_KEYSPACE) + 1) + "";
    }

    public interface Operation {
        void go();
    }

    public static Map<ChosenAction, ArrayList<Long>> getLatencies(int iterations, Map<ChosenAction, Operation> actions) {
        Map<ChosenAction, ArrayList<Long>> latencies = new HashMap<ChosenAction, ArrayList<Long>>();
        for (ChosenAction action : actions.keySet()) {
            latencies.put(action, new ArrayList<Long>());
        }
        
        for (int i = 0; i<iterations; i++) {
            ChosenAction action = chooseAction();
            Operation op = actions.get(action);
            ArrayList<Long> actionLatencies = latencies.get(action);
            addLatency(op, actionLatencies);
        }

        return latencies;
    }

    private static void addLatency(Operation op, ArrayList<Long> latencies) {
        long before = System.nanoTime();
        op.go();
        long after = System.nanoTime();
        latencies.add(after - before);
    }

    // Assumption: latencies is sorted in ascending order
    private static Long percentile(ArrayList<Long> latencies, int percentile) {
        return latencies.get((int) Math.ceil((percentile / 100.0) * latencies.size()));
    }

    private static double stdDeviation(ArrayList<Long> latencies, Double avgLatency) {
        double stdDeviation = latencies.stream()
            .mapToDouble(Long::doubleValue)
            .reduce(0.0, (stdDev, latency) -> stdDev + Math.pow(latency - avgLatency, 2));
        return Math.sqrt(stdDeviation / latencies.size());
    }

    // This has the side-effect of sorting each latencies ArrayList
    public static Map<ChosenAction, LatencyResults> calculateResults(Map<ChosenAction, ArrayList<Long>> actionLatencies) {
        Map<ChosenAction, LatencyResults> results = new HashMap<ChosenAction, LatencyResults>();

        for (Map.Entry<ChosenAction, ArrayList<Long>> entry : actionLatencies.entrySet()) {
            ChosenAction action = entry.getKey();
            ArrayList<Long> latencies = entry.getValue();

            Double avgLatency = latencies
                .stream()
                .collect(Collectors.summingLong(Long::longValue)) / Double.valueOf(latencies.size());

            Collections.sort(latencies);
            results.put(action, new LatencyResults(
                avgLatency,
                percentile(latencies, 50),
                percentile(latencies, 90),
                percentile(latencies, 99),
                stdDeviation(latencies, avgLatency)
            ));
        }

        return results;
    }

    public static void printResults(Map<ChosenAction, LatencyResults> resultsMap) {
        for (Map.Entry<ChosenAction, LatencyResults> entry : resultsMap.entrySet()) {
            ChosenAction action = entry.getKey();
            LatencyResults results = entry.getValue();
            System.out.println(
                "Avg. time in ms per " + action + ": " + results.avgLatency / 1000000.0
            );
            System.out.println(action + " p50 latency in ms: " + results.p50Latency / 1000000.0);
            System.out.println(action + " p90 latency in ms: " + results.p90Latency / 1000000.0);
            System.out.println(action + " p99 latency in ms: " + results.p99Latency / 1000000.0);
            System.out.println(action + " std dev in ms: " + results.stdDeviation / 1000000.0);
        }
    }
}

File: go/benchmarks/benchmarking.go
// Copyright GLIDE-for-Redis Project Contributors - SPDX Identifier: Apache-2.0

package main

import (
	"encoding/json"
	"fmt"
	"log"
	"math"
	"math/rand"
	"os"
	"sort"
	"strings"
	"time"
)

type connectionSettings struct {
	host               string
	port               int
	useTLS             bool
	clusterModeEnabled bool
}

func runBenchmarks(runConfig *runConfiguration) error {
	connSettings := &connectionSettings{
		host:               runConfig.host,
		port:               runConfig.port,
		useTLS:             runConfig.tls,
		clusterModeEnabled: runConfig.clusterModeEnabled,
	}

	err := executeBenchmarks(runConfig, connSettings)
	if err != nil {
		return err
	}

	if runConfig.resultsFile != os.Stdout {
		return writeResults(runConfig.resultsFile)
	}

	return nil
}

type benchmarkConfig struct {
	clientName         string
	numConcurrentTasks int
	clientCount        int
	dataSize           int
	minimal            bool
	connectionSettings *connectionSettings
	resultsFile        *os.File
}

func executeBenchmarks(runConfig *runConfiguration, connectionSettings *connectionSettings) error {
	var benchmarkConfigs []benchmarkConfig
	for _, clientName := range runConfig.clientNames {
		for _, numConcurrentTasks := range runConfig.concurrentTasks {
			for _, clientCount := range runConfig.clientCount {
				for _, dataSize := range runConfig.dataSize {
					benchmarkConfig := benchmarkConfig{
						clientName:         clientName,
						numConcurrentTasks: numConcurrentTasks,
						clientCount:        clientCount,
						dataSize:           dataSize,
						minimal:            runConfig.minimal,
						connectionSettings: connectionSettings,
						resultsFile:        runConfig.resultsFile,
					}

					benchmarkConfigs = append(benchmarkConfigs, benchmarkConfig)
				}
			}
		}

		for _, config := range benchmarkConfigs {
			err := runSingleBenchmark(&config)
			if err != nil {
				return err
			}
		}

		fmt.Println()
	}

	return nil
}

func runSingleBenchmark(config *benchmarkConfig) error {
	fmt.Printf("Running benchmarking for %s client:\n", config.clientName)
	fmt.Printf(
		"\n =====> %s <===== clientCount: %d, concurrentTasks: %d, dataSize: %d \n\n",
		config.clientName,
		config.clientCount,
		config.numConcurrentTasks,
		config.dataSize,
	)

	clients, err := createClients(config)
	if err != nil {
		return err
	}

	benchmarkResult := measureBenchmark(clients, config)
	if config.resultsFile != os.Stdout {
		addJsonResults(config, benchmarkResult)
	}

	printResults(benchmarkResult)
	return closeClients(clients)
}

type benchmarkClient interface {
	connect(connectionSettings *connectionSettings) error
	set(key string, value string) (string, error)
	get(key string) (string, error)
	close() error
	getName() string
}

func createClients(config *benchmarkConfig) ([]benchmarkClient, error) {
	var clients []benchmarkClient
	for clientNum := 0; clientNum < config.clientCount; clientNum++ {
		var client benchmarkClient
		switch config.clientName {
		case goRedis:
			client = &goRedisBenchmarkClient{}
		case glide:
			client = &glideBenchmarkClient{}
		}

		err := client.connect(config.connectionSettings)
		if err != nil {
			return nil, err
		}

		clients = append(clients, client)
	}

	return clients, nil
}

func closeClients(clients []benchmarkClient) error {
	for _, client := range clients {
		err := client.close()
		if err != nil {
			return err
		}
	}

	return nil
}

var jsonResults []map[string]interface{}

func writeResults(file *os.File) error {
	fileInfo, err := file.Stat()
	if err != nil {
		return err
	}

	if fileInfo.Size() != 0 {
		decoder := json.NewDecoder(file)
		var existingData []map[string]interface{}
		err = decoder.Decode(&existingData)
		if err != nil {
			return err
		}

		jsonResults = append(existingData, jsonResults...)
	}

	marshalledJson, err := json.Marshal(jsonResults)
	if err != nil {
		return err
	}
	_, err = file.WriteAt(marshalledJson, 0)
	if err != nil {
		return err
	}

	return nil
}

type benchmarkResults struct {
	iterationsPerTask int
	duration          time.Duration
	tps               float64
	latencyStats      map[string]*latencyStats
}

func measureBenchmark(clients []benchmarkClient, config *benchmarkConfig) *benchmarkResults {
	var iterationsPerTask int
	if config.minimal {
		iterationsPerTask = 1000
	} else {
		iterationsPerTask = int(math.Min(math.Max(1e5, float64(config.numConcurrentTasks*1e4)), 1e7))
	}

	actions := getActions(config.dataSize)
	duration, latencies := runBenchmark(iterationsPerTask, config.numConcurrentTasks, actions, clients)
	tps := calculateTPS(latencies, duration)
	stats := getLatencyStats(latencies)
	return &benchmarkResults{
		iterationsPerTask: iterationsPerTask,
		duration:          duration,
		tps:               tps,
		latencyStats:      stats,
	}
}

func calculateTPS(latencies map[string][]time.Duration, totalDuration time.Duration) float64 {
	numRequests := 0
	for _, durations := range latencies {
		numRequests += len(durations)
	}

	return float64(numRequests) / totalDuration.Seconds()
}

type operations func(client benchmarkClient) (string, error)

const (
	getExisting    = "get_existing"
	getNonExisting = "get_non_existing"
	set            = "set"
)

func getActions(dataSize int) map[string]operations {
	actions := map[string]operations{
		getExisting: func(client benchmarkClient) (string, error) {
			return client.get(keyFromExistingKeyspace())
		},
		getNonExisting: func(client benchmarkClient) (string, error) {
			return client.get(keyFromNewKeyspace())
		},
		set: func(client benchmarkClient) (string, error) {
			return client.set(keyFromExistingKeyspace(), strings.Repeat("0", dataSize))
		},
	}

	return actions
}

const (
	sizeNewKeyspace      = 3750000
	sizeExistingKeyspace = 3000000
)

func keyFromExistingKeyspace() string {
	localRand := rand.New(rand.NewSource(time.Now().UnixNano()))
	return fmt.Sprint(math.Floor(localRand.Float64()*float64(sizeExistingKeyspace)) + 1)
}

func keyFromNewKeyspace() string {
	localRand := rand.New(rand.NewSource(time.Now().UnixNano()))
	totalRange := sizeNewKeyspace - sizeExistingKeyspace
	return fmt.Sprint(math.Floor(localRand.Float64()*float64(totalRange) + sizeExistingKeyspace + 1))
}

type actionLatency struct {
	action  string
	latency time.Duration
}

func runBenchmark(
	iterationsPerTask int,
	concurrentTasks int,
	actions map[string]operations,
	clients []benchmarkClient,
) (totalDuration time.Duration, latencies map[string][]time.Duration) {
	latencies = map[string][]time.Duration{
		getExisting:    {},
		getNonExisting: {},
		set:            {},
	}

	numResults := concurrentTasks * iterationsPerTask
	results := make(chan *actionLatency, numResults)
	start := time.Now()
	for i := 0; i < concurrentTasks; i++ {
		go runTask(results, iterationsPerTask, actions, clients)
	}

	for i := 0; i < numResults; i++ {
		result := <-results
		latencies[result.action] = append(latencies[result.action], result.latency)
	}

	return time.Since(start), latencies
}

func runTask(results chan<- *actionLatency, iterations int, actions map[string]operations, clients []benchmarkClient) {
	for i := 0; i < iterations; i++ {
		clientIndex := i % len(clients)
		action := randomAction()
		operation := actions[action]
		latency := measureOperation(operation, clients[clientIndex])
		results <- &actionLatency{action: action, latency: latency}
	}
}

func measureOperation(operation operations, client benchmarkClient) time.Duration {
	start := time.Now()
	_, err := operation(client)
	duration := time.Since(start)
	if err != nil {
		log.Print("Error while executing operation: ", err)
	}

	return duration
}

const (
	probGet            = 0.8
	probGetExistingKey = 0.8
)

func randomAction() string {
	localRand := rand.New(rand.NewSource(time.Now().UnixNano()))
	if localRand.Float64() > probGet {
		return set
	}

	if localRand.Float64() > probGetExistingKey {
		return getNonExisting
	}

	return getExisting
}

type latencyStats struct {
	avgLatency   time.Duration
	p50Latency   time.Duration
	p90Latency   time.Duration
	p99Latency   time.Duration
	stdDeviation time.Duration
	numRequests  int
}

func getLatencyStats(actionLatencies map[string][]time.Duration) map[string]*latencyStats {
	results := make(map[string]*latencyStats)

	for action, latencies := range actionLatencies {
		sort.Slice(latencies, func(i, j int) bool {
			return latencies[i] < latencies[j]
		})

		results[action] = &latencyStats{
			// TODO: Replace with a stats library, eg https://pkg.go.dev/github.com/montanaflynn/stats
			avgLatency:   average(latencies),
			p50Latency:   percentile(latencies, 50),
			p90Latency:   percentile(latencies, 90),
			p99Latency:   percentile(latencies, 99),
			stdDeviation: standardDeviation(latencies),
			numRequests:  len(latencies),
		}
	}

	return results
}

func average(observations []time.Duration) time.Duration {
	var sumNano int64 = 0
	for _, observation := range observations {
		sumNano += observation.Nanoseconds()
	}

	avgNano := sumNano / int64(len(observations))
	return time.Duration(avgNano)
}

func percentile(observations []time.Duration, p float64) time.Duration {
	N := float64(len(observations))
	n := (N-1)*p/100 + 1

	if n == 1.0 {
		return observations[0]
	} else if n == N {
		return observations[int(N)-1]
	}

	k := int(n)
	d := n - float64(k)
	interpolatedValue := float64(observations[k-1]) + d*(float64(observations[k])-float64(observations[k-1]))
	return time.Duration(int64(math.Round(interpolatedValue)))
}

func standardDeviation(observations []time.Duration) time.Duration {
	var sum, mean, sd float64
	numObservations := len(observations)
	for i := 0; i < numObservations; i++ {
		sum += float64(observations[i])
	}

	mean = sum / float64(numObservations)
	for j := 0; j < numObservations; j++ {
		sd += math.Pow(float64(observations[j])-mean, 2)
	}

	sd = math.Sqrt(sd / float64(numObservations))
	return time.Duration(sd)
}

func printResults(results *benchmarkResults) {
	fmt.Printf("Runtime (sec): %.3f\n", results.duration.Seconds())
	fmt.Printf("Iterations: %d\n", results.iterationsPerTask)
	fmt.Printf("TPS: %d\n", int(results.tps))

	var totalRequests int
	for action, latencyStat := range results.latencyStats {
		fmt.Printf("===> %s <===\n", action)
		fmt.Printf("avg. latency (ms): %.3f\n", latencyStat.avgLatency.Seconds()*1000)
		fmt.Printf("std dev (ms): %.3f\n", latencyStat.stdDeviation.Seconds()*1000)
		fmt.Printf("p50 latency (ms): %.3f\n", latencyStat.p50Latency.Seconds()*1000)
		fmt.Printf("p90 latency (ms): %.3f\n", latencyStat.p90Latency.Seconds()*1000)
		fmt.Printf("p99 latency (ms): %.3f\n", latencyStat.p99Latency.Seconds()*1000)
		fmt.Printf("Number of requests: %d\n", latencyStat.numRequests)
		totalRequests += latencyStat.numRequests
	}

	fmt.Printf("Total requests: %d\n", totalRequests)
}

func addJsonResults(config *benchmarkConfig, results *benchmarkResults) {
	jsonResult := make(map[string]interface{})

	jsonResult["client"] = config.clientName
	jsonResult["is_cluster"] = config.connectionSettings.clusterModeEnabled
	jsonResult["num_of_tasks"] = config.numConcurrentTasks
	jsonResult["data_size"] = config.dataSize
	jsonResult["client_count"] = config.clientCount
	jsonResult["tps"] = results.tps

	for key, value := range results.latencyStats {
		jsonResult[key+"_p50_latency"] = value.p50Latency.Seconds() * 1000
		jsonResult[key+"_p90_latency"] = value.p90Latency.Seconds() * 1000
		jsonResult[key+"_p99_latency"] = value.p99Latency.Seconds() * 1000
		jsonResult[key+"_average_latency"] = value.avgLatency.Seconds() * 1000
		jsonResult[key+"_std_dev"] = value.stdDeviation.Seconds() * 1000
	}

	jsonResults = append(jsonResults, jsonResult)
}
